### PR TITLE
Micrsoft.DelegatedNetwork RP RestAPI spec

### PR DESCRIPTION
**Pull request by tamilmani1989**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by chiragg4u** [comment]
Please go through the contribution checklist added as part of the review and mark them completed.

**Comment by tamilmani1989** [comment]
> Please go through the contribution checklist added as part of the review and mark them completed.

done

**Comment by chiragg4u** [comment]
I see you've created Preview swagger version under stable path normal pattern is to put it under preview path. Is there any specific reason?
Also for completely new RP's we are recommending to look at RPaaS (RP as a service). Have you evaluated it?
Otherwise changes looks good from ARM API design. SignedOff from ARM side, please follow the process given @ https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/202/Overall-Process-of-AME-Onboarding

**Comment by tamilmani1989** [comment]
> I see you've created Preview swagger version under stable path normal pattern is to put it under preview path. Is there any specific reason?
> Also for completely new RP's we are recommending to look at RPaaS (RP as a service). Have you evaluated it?
> Otherwise changes looks good from ARM API design. SignedOff from ARM side, please follow the process given @ https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/202/Overall-Process-of-AME-Onboarding

done

**Review by PhoenixHe-NV — APPROVED** [review]

**Comment by NullMDR** [comment]
@tamilmani1989 Ready to merge?

**Comment by NullMDR** [comment]
@sanjaiganesh Does this new RP need to use RPaaS?

**Comment by tamilmani1989** [comment]
> @sanjaiganesh Does this new RP need to use RPaaS?

That option would not work for us and we already evaluated Rpaas. Also its not GA yet. We started implementing RP already.

**Comment by tamilmani1989** [comment]
@NullMDR sorry i thought of deleting reply. i deleted entire comment. Lets hold on this PR, identified few changes to spec lately. will push that and re-review this

**Comment by NullMDR** [comment]
@tamilmani1989 Ping me when you are ready.

File: specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json [hunk 1]
```diff
@@ -0,0 +1,820 @@
...
+      "type": "oauth2",
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
+      "flow": "implicit",
+      "description": "Azure Active Directory OAuth2 Flow",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
```
Comment (ravbhatnagar) [inline specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:36]: missing Patch which is a requirement

File: specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json [hunk 2]
```diff
@@ -0,0 +1,820 @@
...
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "name": "delegatedSubnetParameters",
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/DelegatedSubnetParameters"
```
Comment (ravbhatnagar) [inline specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:255]: request body and response body must use the same model.

File: specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json [hunk 3]
```diff
@@ -0,0 +1,820 @@
...
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponse"
+            }
+          }
+        },
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/virtualNetworks/{vnetName}/subnets/{subnetName}/providers/Microsoft.DelegatedNetwork/delegatedSubnet": {
```
Comment (ravbhatnagar) [inline specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:179]: this path wont work with ARM. You need a {name} segment to indicate the name of the delegatedSubnet

Comment (ravbhatnagar) [reply specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:179]: and then add a GET api to list all delegatedSubnets under a subnet.

Comment (ravbhatnagar) [reply specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:179]: pluralize the resource type name so delegatedSubnets

**Review by ravbhatnagar — COMMENTED** [review]
@tamilmani1989 - adding comments for delegatedSUbnet APIs only, since the controller type was already reviewed by Chirag and Marcin.

File: specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json (hunk 3 above)
Comment (tamilmani1989) [reply specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:179]: > this path wont work with ARM. You need a {name} segment to indicate the name of the delegatedSubnet

can we give same subnet name?

Comment (tamilmani1989) [reply specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:179]: > and then add a GET api to list all delegatedSubnets under a subnet.

subnet is same here and the only thing is its delegated. so there won't be any list of delegated subnets under subnet

File: specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json (hunk 2 above)
Comment (tamilmani1989) [reply specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:255]: why is that requirement? response body can have state saying succeeded or failed and request body doesn't need that.

Comment (tamilmani1989) [reply specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:255]: i have seen many existing RPs using different request and response body

**Review by ravbhatnagar — COMMENTED** [review]
looks good, signing off from ARM side.

**Comment by tamilmani1989** [comment]
@NullMDR its good to merge now

File: specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json [hunk 4]
```diff
@@ -0,0 +1,1037 @@
...
+            "$ref": "#/definitions/DelegatedController"
+          },
+          "description": "An array of Delegated controller resources."
+        }
+      },
+      "required": [
+        "value"
+      ]
+    },
+    "DelegatedSubnets": {
```
Comment (xccc-msft) [inline specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json:799]: For Java SDK automation error, you may consider to rename this or change the operation ids.
Potential suggestion:
`DelegatedSubnets` -> `DelegatedSubnetList`/`DelegatedSubnetItems`
or
`DelegatedSubnet(s)_OPERATION` -> `DelegatedSubnet(s)Service_OPERATION`

Comment (tamilmani1989) [reply specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/DelegatedNetwork.json]: done

**Pull request merged**